keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,205 @@
+
+## Motivation
+
+Recent operator development has shown that complex operators often depend on other operators to function. One workaround commonly seen can be found in the [Flink Demo](https://github.com/kudobuilder/operators/tree/master/repository/flink/docs/demo/financial-fraud). It requires manual installation of the dependency operators while skipping the instance creation (using the `--skip-instance` option). The `Instance` is then "manually" created during the execution of the `deploy` plan. This KEP is aiming at improving this experience, streamlining the developer and user experience.
+
+### Goals
+
+Dependencies can be a complex topic. This KEP is not trying to boil the dependency ocean but rather limits itself to installation dependencies only.
+
+### Non-Goals
+
+Dependency on an already running `Instance` is a non-goal. It is easy to imagine a situation when a new operator (e.g Kafka) may want to depend on the existing Zookeeper instance. However, such life-cycle dependency presents major challenges e.g. what happens when Zookeeper is removed? What happens when Zookeeper is upgraded and the new version is incompatible with the current Kafka `Instance`? How can compatibility be ensured? This KEP deliberately ignores this area and instead focuses on installation dependencies. Additionally, this KEP does not address output variables or referencing `Instance` resources.
+
+## Proposal
+
+KUDO operators **already have** a mechanism to deal with installation dependencies called [plans, phases, and steps](https://kudo.dev/docs/developing-operators/plans.html#overview) with serial or parallel execution strategy. This mechanism is already powerful enough to express any dependency hierarchy including transitive dependencies (see more about this in [implementation details](#implementation-details)). The core of this proposal is to reuse this mechanism and extend it with the ability to install operators.
+
+### Implementation details
+
+#### Operator Task
+
+A new task kind `Operator` is introduced which extends `operator.yaml` with the ability to install dependencies. Let's take a look at the Kafka+Zookeeper example:
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+name: "kafka"
+operatorVersion: "1.3.1"
+kudoVersion: 0.12.0
+kubernetesVersion: 1.14.8
+appVersion: 2.5.0
+url: https://kafka.apache.org/
+tasks:
+  - name: zookeeper-operator
+    kind: Operator
+    spec:
+        package: zookeeper
+
+  ...
+
+plans:
+  deploy:
+    strategy: serial
+    phases:
+      - name: deploy-zookeeper
+        strategy: serial
+        steps:
+          - name: dummy
+            tasks:
+              - zookeeper-operator
+      - name: deploy-kafka
+        strategy: serial
+        steps:
+          ...
+```
+
+The `zookeeper-operator` task specification is equivalent to `kudo install zookeeper` CLI command which installs the Zookeeper package from the official repo. Here is a complete `Operator` task specification:
```

Comment:
what if instance with that name alrrady exists? Will it continue installing or fail? If it continues, then it's basically reusing already running instance which is a non-goal I understand...

Reply:
> what if instance with that name alrrady exists? 

this can't happen, right? you can't have two instances [with the same name](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/), the same way you can't have two pods with the same name (or is it per namespace? 🤔)

Reply:
@zen-dog it depends on whether you're using an idempotent `create-or-update` or non-idempotent `create-or-fail`, right?

Reply:
For `Instance`s we use `create-or-fail`, otherwise, we risk changing an already existing instance for some other operator. These problem falls into _"depending on an existing Instance"_ category and is out of scope for this KEP.